Add renderer for Slideshow content

nucleogenesis opened the pull request:
<!--
 1. Following guidance below, replace …'s with your own words
 2. After saving the PR, tick of completed checklist items
 3. Skip checklist items that are not applicable or not necessary
 4. Delete instruction/comment blocks
-->

### Summary
<!--
 * description of the change
 * manual verification steps performed
 * screenshots if the PR affects the UI
-->

This adds all functionality necessary for importing a Slideshow content kind from a Studio instance and displaying the Slideshow via the new slideshow_renderer plugin.
…

### Reviewer guidance
<!--
 * how can a reviewer test these changes?
 * are there any risky areas that deserve extra testing
-->

1. Once you have this branch running locally, go to Device and click [Import].
2. Add `http://nucleogenesis-slideshow-content-kind.studio.cd.learningequality.org` as a "Local network or internet" source.
3. Import from that source.
4. You should see "Learning Equality Slideshow Demo" as an available content node. Import that.
5. Find that content in Learn and test.

### Contributor Checklist

PR process:

- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

Testing:

- [x] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [x] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests

### Reviewer Checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

nucleogenesis committed:
Slideshow base renderer

nucleogenesis committed:
refitting for updated le-utils

nucleogenesis committed:
require latest le-utils in requirements/base

nucleogenesis committed:
updated pipfile.lock le-utils version

nucleogenesis committed:
Revert "updated pipfile.lock le-utils version"

This reverts commit f4d6544d622f7b8c89c81d0dc08004a6d8831840.

nucleogenesis committed:
pagination & caption style updates

nucleogenesis committed:
better responsive images in img tags

nucleogenesis committed:
use more vertical screen real estate

nucleogenesis committed:
Binding the alt tag on slide images

nucleogenesis committed:
a11y updates to slide images

rtibbles commented:
Note that since you made the database migration for this, a new one has been added to develop - best to delete your migration and rerun `makemigrations`.

Also, the rebase seems to have done something most peculiar?

nucleogenesis committed:
fixed formatting and migration conflict

indirectlylit commented:
(updated title b/c WIP is in the label)

nucleogenesis committed:
add icon for slideshow kind

nucleogenesis commented:
@rtibbles @indirectlylit 

Would you be willing to give this a review so we can get it merged?

A working demo of this with a Slideshow content is available on a demo server here:

(login: devowner / admin123)
http://35.229.70.110
http://35.229.70.110/learn/#/topics/c/da16403d0e22582c8ab589dc4e4fce9c  -> The slideshow content directly

rtibbles commented:
Yeah, now that we've branched for 0.12.y, we can merge this to develop sooner!

indirectlylit commented on kolibri/core/content/migrations/0017_auto_20190415_1855.py:
```diff
@@ -1,19 +0,0 @@
-        migrations.AlterField(
-            model_name="channelmetadata",
-            name="published_size",
-            field=models.BigIntegerField(blank=True, default=0, null=True),
```
We should generally just create new migrations rather than removing and re-creating existing ones. This ensures that every DB out there has a path 'upward'.

Also, we're not consistent about it, but it's good to name the migration something meaningful like `0018_do_something_special.py`

nucleogenesis replied:
@indirectlylit that makes sense. I cannot recall exactly why I deleted this migration. Should I leave it deleted in this instance since the repo includes a migration file that replaces this one?

rtibbles replied:
The other thing to bear in mind is that other migrations might be added which need to be previous to this in the migration tree (e.g. https://github.com/learningequality/kolibri/pull/5497/files#diff-21bfbbd534842850877c325a17b3d32a), so the correct migration (particularly on the develop branch) can end up being a bit of a moving target.

The best thing to do in this case would be to reinstate the previous `0017` migration, then create a new migration just for what you are doing.

If the migration I note above ends up getting merged before your PR, then we will need to recreate the migration to take that into account. Otherwise it can get sorted out by the person who merges it into develop.

indirectlylit commented on kolibri/plugins/slideshow_render/assets/src/views/SlideshowRendererComponent.vue:
```diff
@@ -0,0 +1,214 @@
+    position: absolute;
+    top: 8px;
+    right: 8px;
+    z-index: 100;
```
z-indexes should generally follow Material's [default elevations](https://material.io/design/environment/elevation.html#default-elevations)

indirectlylit commented on kolibri/plugins/slideshow_render/assets/src/views/SlideshowRendererComponent.vue:
```diff
@@ -0,0 +1,214 @@
+      manifest: {},
+      slides: null,
+      currentSlide: null,
+      hooperSettings: {
```
I don't actually see this used anywhere

indirectlylit commented on kolibri/plugins/slideshow_render/assets/src/views/SlideshowRendererComponent.vue:
```diff
@@ -0,0 +1,214 @@
+      },
+    }),
+    computed: {
+      slideshow_images: function() {
```
we conventionally use `camelCase` on the JS side and `snake_case` on the python side

indirectlylit commented on kolibri/plugins/slideshow_render/assets/src/views/SlideshowRendererComponent.vue:
```diff
@@ -0,0 +1,214 @@
+    computed: {
+      slideshow_images: function() {
+        const files = this.files;
+        return files.filter(file => file.preset != 'Slideshow Manifest');
```
We should use programmatic constant strings here like `slideshow_manifest` rather than things that look human-readable.

Otherwise it becomes tempting to display it to the end-user, like in #2371

indirectlylit commented on kolibri/plugins/slideshow_render/assets/src/views/SlideshowRendererComponent.vue:
```diff
@@ -0,0 +1,214 @@
+
+<script>
+
+  import _ from 'lodash';
```
should be written `import orderBy from 'lodash/orderBy';` to avoid pulling in the entire lodash library

indirectlylit commented on kolibri/plugins/slideshow_render/assets/src/views/SlideshowRendererComponent.vue:
```diff
@@ -0,0 +1,214 @@
+      fetch(this.defaultFile.storage_url, { method: 'GET' })
+        .then(response => {
+          return response.status === 200 ? response.json() : {};
+        })
```
we will probably need some loading and error states?

see e.g. `handleApiError` and `CORE_SET_PAGE_LOADING`

nucleogenesis replied:
@indirectlylit 

I refactored it to just return `response.json()` and added a `catch` block which dispatches `handleApiError` with the error object.

As for the `CORE_SET_PAGE_LOADING` - I tried:

```
mounted() {
      /*
        Using the manifest file, get the JSON from the manifest, then
        use the manifest JSON to get all slide images and metadata.
      */
      this.$store.commit("CORE_SET_PAGE_LOADING", true);
      fetch(this.defaultFile.storage_url, { method: 'GET' })
        .then(response => {
          return response.json();
        })
        .then(manifest_data => {
          this.manifest = manifest_data.slideshow_data;

          this.slides = orderBy(
            this.manifest.map(image => {
              return {
                storage_url: this.slideshowImages.find(
                  sFile => checksumFromFile(sFile) == image.checksum
                ).storage_url,
                caption: image.caption,
                sort_order: image.sort_order,
                id: image.checksum,
                descriptive_text: image.descriptive_text,
              };
            }),
            ['sort_order'],
            ['asc']
          );
          this.$store.commit("CORE_SET_PAGE_LOADING", false);
        })
        .catch(error => {
          this.$store.commit("CORE_SET_PAGE_LOADING", false);
          this.$store.dispatch('handleApiError', error);
        });
```

For some reason, this infinitely loops calling the `mounted()` method over and over. I looked a bit into that mutation and I'm not seeing how or why that would re-mount the component. 

Any thoughts on why this might be the case?

Also - FWIW - the functionality of the slideshow is to show a spinning loader until the images are all loaded. I've tested with bandwidth throttling to slow down the networking and it showed the spinner until it was fully loaded and ready.

rtibbles replied:
Sorry, should have caught this sooner - you should use the kolibri `client` module: `import client from 'kolibri.client';` to do any fetches (rather than using the `fetch` API which is not as widely supported: https://caniuse.com/#feat=fetch (e.g. IE11, but also many others).

nucleogenesis replied:
@rtibbles thanks for letting me know! Done and pushed.

indirectlylit commented on kolibri/plugins/slideshow_render/assets/src/views/SlideshowRendererComponent.vue:
```diff
@@ -0,0 +1,214 @@
+                  sFile =>
+                    sFile.storage_url
+                      .split('/')
+                      [sFile.storage_url.split('/').length - 1].split('.')[0] == image.checksum
```
somewhat complex string logic .. maybe define this as a helper method with a unit test

indirectlylit commented on kolibri/plugins/slideshow_render/assets/src/views/SlideshowRendererComponent.vue:
```diff
@@ -0,0 +1,214 @@
+        this.currentSlide = this.slides[payload.currentSlide];
+      },
+      slideTextId(id) {
+        return 'descriptive-text-' + id;
```
If for some reason we ever wanted to show two slide shows at the same time, we'd need to make this more unique. For example, 

```javascript
let component_id_counter = 0;

export default {
  beforeCreate() {
    this.component_id = component_id_counter.toString();
    component_id_counter += 1;
  },
  methods: {
    slideTextId(id) {
      return `${component_id}-descriptive-text-${id}`;
    }
  }
};
```

indirectlylit commented on kolibri/plugins/slideshow_render/assets/src/views/SlideshowRendererComponent.vue:
```diff
@@ -0,0 +1,214 @@
+    right: 8px;
+    z-index: 100;
+    fill: white;
+  }
```
suuuper minor but most other places in the code we put a line break between class definitions

indirectlylit commented on kolibri/plugins/slideshow_render/assets/src/views/SlideshowRendererComponent.vue:
```diff
@@ -0,0 +1,214 @@
+  }
+  .slideshow-renderer {
+    position: relative;
+    height: 70vh;
```
we've tended to avoid the `vh` and `vw` units due to [early android browsers](https://caniuse.com/#feat=viewport-units)

indirectlylit commented on kolibri/plugins/slideshow_render/assets/src/views/SlideshowRendererComponent.vue:
```diff
@@ -0,0 +1,214 @@
+
+  import themeMixin from 'kolibri.coreVue.mixins.themeMixin';
+  import contentRendererMixin from 'kolibri.coreVue.mixins.contentRendererMixin';
+  import responsiveElement from 'kolibri.coreVue.mixins.responsiveElement';
```
when possible it's nice to avoid reliance on `responsiveElement` but sometimes it's a necessary evil

nucleogenesis replied:
Noted! 👍 Thanks Devon

indirectlylit commented on kolibri/plugins/slideshow_render/assets/src/views/SlideshowRendererComponent.vue:
```diff
@@ -0,0 +1,214 @@
+      },
+      files: {
+        type: Array,
+        default: () => [],
```
whenever props are complex types (objects or arrays) we should also provide a `validator` function

nucleogenesis replied:
Thanks for letting me know - I've added validators to both props. Do you think they cover what I'll need to validate well enough?

indirectlylit replied:
I guess I'm actually still left a bit confused as to what form of data these props expect. From the validators I see a schema like:

```
defaultFile: { storage_url: "*json" }
```

and 

```
files: [
  { storage_url: "*" },
  //...
]
```

but then down below I also see a reference to `file.preset`.

I guess what I would expect to see in either comments, validator function, or both is a complete description of the shape of the data for the things that the component is calling a "file" (or maybe more accurately a "file descriptor object"?).

indirectlylit commented on kolibri/plugins/slideshow_render/assets/src/views/SlideshowRendererComponent.vue:
```diff
@@ -0,0 +1,214 @@
+            :aria-labelledby="slideTextId(slide.id)"
+          >
+        </div>
+        <div :id="slideTextId(slide.id)" class="hidden-descriptive-text">
```
we have a global `visuallyhidden` class which should achieve this

indirectlylit commented on kolibri/plugins/slideshow_render/assets/src/views/SlideshowRendererComponent.vue:
```diff
@@ -0,0 +1,214 @@
+    Navigation as HooperNavigation,
+    Pagination as HooperPagination,
+  } from 'hooper';
+  import '../../styles/custom-hooper.css';
```
let's move `custom-hooper.css` to the same directory as this component, since `SlideshowRendererComponent.vue` is the only place `Hooper` is used.

Also note that importing the css file like this will drop the styles into the global scope.

I don't know for sure, but we might be able scope it using something like...

```html
<style scoped>
  /deep/ .slideshow-renderer {
    @import './custom-hooper.css';
  }
</style>
```

or some other variation of sass-y goodness

nucleogenesis replied:
I refactored to import using CSS and moved the file. However I had issues when trying to import the CSS using `/deep/`  - since I have `<style scoped>` shouldn't that ensure the Hooper CSS is only used on this component?

indirectlylit replied:
yes if it works without `/deep/` that's great. Sometimes I've encountered issues where that wasn't the case

indirectlylit commented on kolibri/plugins/slideshow_render/assets/styles/custom-hooper.css:
```diff
@@ -0,0 +1,145 @@
+  transform: translateX(50%);
+}
+.hooper-indicators {
+  display: flex;
```
we should check whether things still work without flexbox since older android browsers don't support

nucleogenesis replied:
@indirectlylit I imagine that it could possibly be made to work without flexbox. Non-active slides are hidden to the side and the position of the active element is placed into view. I wish I'd realized that this wasn't going to work previously I recall now having read that we can't use flex in the style guide.

I'll see what I can do to replicate the effect without flexbox but may opt to try out other OSS plugins that give similar features if I can find one that doesn't make use of flexbox.

indirectlylit replied:
see a similar thread that just occurred with @bjester :)

https://github.com/learningequality/kolibri/pull/5475#discussion_r283904383

if it's too hard to change for now, let's leave it and just make sure we're aware of it and give it extra attention during integration testing

nucleogenesis replied:
I think I managed a solid solution. A good point from @bjester is 

>  I think in these scenarios where there is a question of whether this will work correctly, you can take into account the use and what happens when it breaks.

In my case it broke everything - Hooper was very reliant on flex to manage the spacing and flow of the slides.

indirectlylit commented on kolibri/plugins/slideshow_render/assets/styles/custom-hooper.css:
```diff
@@ -0,0 +1,145 @@
+}
+.hooper-indicator.is-active,
+.hooper-indicator:hover {
+  /* background-color: #4285f4; */
```
dead code

indirectlylit reviewed (commented):
haven't tested yet, but here are some small notes - nothing major

bjester commented on kolibri/plugins/slideshow_render/assets/src/views/SlideshowRendererComponent.vue:
```diff
@@ -0,0 +1,214 @@
+    background-size: contain;
+
+    img {
+      object-fit: contain;
```
If you don't already, I think you'll need a polyfill for this. IE 11 and old Android browsers do not support `object-fit`

nucleogenesis replied:
Thanks @bjester - I hadn't considered how new object-fit is - I found and implemented a simple and working polyfill. In testing I was also was reminded that IE11 doesn't support fetch() so had to revise that to use XMLHTTPRequest.

bjester replied:
Nice! Yeah, object-fit is really nice so glad the polyfill worked!

nucleogenesis committed:
refactor to make Hooper not use flexbox

nucleogenesis committed:
css optimization and style fixes, formatting fixes

nucleogenesis committed:
refactor and test checksum extraction, style fixes

nucleogenesis committed:
add validators for props

nucleogenesis committed:
use constant for comparing presets

nucleogenesis committed:
import hooper css in scoped sass on component, moved style file

nucleogenesis committed:
dispatch handleApiError in manifest fetch

nucleogenesis committed:
polyfill for object-fit:contain, replace fetch with XMLHTTPRequest

nucleogenesis committed:
fetch using kolibri.client

nucleogenesis commented:
@indirectlylit @rtibbles @bjester 

I've gone through all of the feedback above and pushed several commits addressing the issues, particularly those that could affect compatibility.

- Refactored Hooper CSS to work without flexbox
- Replaced my `fetch()` with `kolibri.client`
- Use a polyfill for the CSS `object-fit` to accommodate IE11 and older Android.
- Misc refactoring, code cleanup, formatting fixes.

I've tested in Chrome & Firefox on Linux, IE11 on Windows 7 (Browser Stack) and 10 (Proper Install) and Chrome on Android v4 (Browser Stack).

I have also updated the "Reviewer Guidelines" in the description for the PR which will help you quickly get Slideshow content into your local Kolibri instance and ready for test. Please let me know if you run into any issues!

indirectlylit commented:
Hey @nucleogenesis  -

I pushed a quick commit to resolve some minor conflicts. Let's just follow up on the [migrations thread](https://github.com/learningequality/kolibri/pull/5391#discussion_r283567537) above, and then we can merge this in and iterate in a future PR

nucleogenesis committed:
gherkin story for slideshow

nucleogenesis committed:
fix gherkin story spelling, indentation - clarify some stories

nucleogenesis committed:
reinstate previous migration tree

nucleogenesis committed:
update gherkin action descriptions

nucleogenesis committed:
Slideshow options on content node migration

nucleogenesis commented on integration_testing/features/learner/learner-engage-slideshow.feature:
```diff
@@ -0,0 +1,59 @@
+Feature: Learner engages with content of the slideshow kind
```
@radinamatic Would you be able to give this gherkin story a look and let me know if it covers everything you think that should be covered in a story for the slideshow?

radinamatic commented on integration_testing/features/learner/learner-engage-slideshow.feature:
```diff
@@ -0,0 +1,59 @@
+        And I see left and right arrow buttons on the left and right sides of the screen
+        And I see the exit full screen button in the top right corner of the screen
+        And I can navigate the content by keyboard, swiping or clicking the left and right arrow buttons
+      When I am viewing in full screen mode and click the exit full screen button or hit ESC
```
It doesn't seem necessary to repeat you are in full screen mode here, shorten to `When I click the exit full screen button or hit ESC`.

radinamatic commented on integration_testing/features/learner/learner-engage-slideshow.feature:
```diff
@@ -0,0 +1,59 @@
+      When I view a slide that has caption text
+      Then I see the caption text beneath the slide image
+      When I am viewing the first slide
+      Then I the left-most pagination dot is filled in with color and the rest of them are white
```
Delete the first `I`.

radinamatic commented on integration_testing/features/learner/learner-engage-slideshow.feature:
```diff
@@ -0,0 +1,59 @@
+        And tapping the left key on my keyboard does nothing
+        And swiping the image to the right does nothing
+      When I am viewing the last slide
+      Then I the right-most pagination dot is filled in with color and the rest of them are white
```
Delete the first `I`.

radinamatic reviewed (changes requested):
@nucleogenesis I pulled your branch locally but wasn't able to add `http://nucleogenesis-slideshow-content-kind.studio.cd.learningequality.org` as a "Local network or internet" source and do a proper test. 

Gherkin scenario looks complete otherwise, I left just a couple of comments. :+1:

nucleogenesis committed:
Gherkin grammar fixes

radinamatic commented:
Ok, a few suggestions for improvements related to keyboard navigation, feel free to work on them in the following PR, if preferred:

1. Make the visual focus indicator around the full screen button bigger - at the moment it is overlapping the button itself, and it's not easily noticeable.
2. When you navigate the dots with the <kbd>Tab</kbd> key, but then you switch to using arrow keys, these seem to take "precedence" (that is, slides become active as keys are pressed), but the focus and the ring remain where the <kbd>Tab</kbd> key stopped previously, and that got me confused for a moment. I don't know if it would be possible, or even advisable, to make the focus and the ring jump to  the active slide and dot too?

![slideshow](https://user-images.githubusercontent.com/1457929/61423515-5a3c8f80-a910-11e9-8d9b-15fe8edab854.gif)

nucleogenesis committed:
Add theme color for slideshow

nucleogenesis commented:
@indirectlylit @radinamatic 

I have opened a PR for the user docs here: https://github.com/learningequality/kolibri-docs/pull/113

nucleogenesis commented:
> Ok, a few suggestions for improvements related to keyboard navigation, feel free to work on them in the following PR, if preferred:
> 
> 1. Make the visual focus indicator around the full screen button bigger - at the moment it is overlapping the button itself, and it's not easily noticeable.
> 2. When you navigate the dots with the Tab key, but then you switch to using arrow keys, these seem to take "precedence" (that is, slides become active as keys are pressed), but the focus and the ring remain where the Tab key stopped previously, and that got me confused for a moment. I don't know if it would be possible, or even advisable, to make the focus and the ring jump to  the active slide and dot too?
> 
> ![slideshow](https://user-images.githubusercontent.com/1457929/61423515-5a3c8f80-a910-11e9-8d9b-15fe8edab854.gif)

RE: #1 - I checked this on a different content type (PDF) and this seems to be an issue that could be addressed separately from this PR - perhaps we should create a new issue?

RE #2 - That is interesting. I wonder what the ideal solution would be for people who are most inclined to need to use keyboard navigation. To me - FWIW - when I think of the keyboard navigation "focus" functionality, I think of my "focus" indicator as sort of like "where my mouse would be if I were using one". So I wouldn't necessarily expect that focus indicator to also then map to the state of the content that I'm viewing. Rather - I'd consider it to be where hitting the `Enter` key would simulate clicking. In any case - if adjusting how the focus is managed would better serve the most common users of keyboard navigation, then I think that I can make it work accordingly.

nucleogenesis committed:
Add sldieshow tooltip

nucleogenesis committed:
Fix yarn.lock & add string for slideshow
